File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/product/ProductDetailsActivity.java
package com.mlsdev.mlsdevstore.presentaion.product;

import androidx.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.Observable;
import androidx.databinding.ObservableField;
import android.os.Bundle;

import com.mlsdev.mlsdevstore.R;
import com.mlsdev.mlsdevstore.data.model.item.Image;
import com.mlsdev.mlsdevstore.databinding.ActivityProductDetailsBinding;
import com.mlsdev.mlsdevstore.presentaion.BaseActivity;
import com.mlsdev.mlsdevstore.presentaion.adapter.ProductGalleryAdapter;
import com.mlsdev.mlsdevstore.presentaion.utils.CustomObservableBoolean;
import com.mlsdev.mlsdevstore.presentaion.utils.ExtrasKeys;

import java.util.List;

import javax.inject.Inject;

public class ProductDetailsActivity extends BaseActivity {
    private ActivityProductDetailsBinding binding;
    private ProductDetailsViewModel viewModel;

    @Inject
    ProductGalleryAdapter galleryAdapter;

    public ProductDetailsViewModel getViewModel() {
        return viewModel;
    }

    public static void launch(Context context, Bundle bundleProductData) {
        Intent intent = new Intent(context, ProductDetailsActivity.class);
        intent.putExtras(bundleProductData);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_product_details);
        viewModel = ViewModelProviders.of(this, getViewModelFactory()).get(ProductDetailsViewModel.class);
        getErrorInViewHandler().subscribeAllErrorCallbacks(viewModel, true);
        binding.setViewModel(viewModel);
        viewModel.setProductDetailsData(getIntent().getExtras());
        viewModel.descriptionIsDisplayed.addOnPropertyChangedCallback(descriptionCallBack);
        viewModel.imageList.addOnPropertyChangedCallback(galleryImagesLoadedCallback);
        binding.productImagesGallery.setAdapter(galleryAdapter);
        getLifecycle().addObserver(viewModel);
        initToolbar(binding.toolbar);
        displayBackArrow(true);
    }

    private Observable.OnPropertyChangedCallback descriptionCallBack = new Observable.OnPropertyChangedCallback() {
        @Override
        public void onPropertyChanged(Observable observable, int i) {
            if (observable instanceof CustomObservableBoolean && ((CustomObservableBoolean) observable).get()) {
                Bundle data = new Bundle();
                data.putString(ExtrasKeys.KEY_PRODUCT_DESCRIPTION, viewModel.description.get());
                DescriptionBottomSheetFragment bottomSheet = new DescriptionBottomSheetFragment();
                bottomSheet.setArguments(data);
                bottomSheet.show(getSupportFragmentManager(), bottomSheet.getTag());
            }
        }
    };

    private Observable.OnPropertyChangedCallback galleryImagesLoadedCallback = new Observable.OnPropertyChangedCallback() {
        @Override
        public void onPropertyChanged(Observable observable, int i) {
            if (ObservableField.class.isAssignableFrom(observable.getClass())) {
                ObservableField<List<Image>> listObservableField = (ObservableField<List<Image>>) observable;
                galleryAdapter.setImages(listObservableField.get());
            }
        }
    };

    @Override
    protected void onDestroy() {
        viewModel.descriptionIsDisplayed.removeOnPropertyChangedCallback(descriptionCallBack);
        viewModel.imageList.removeOnPropertyChangedCallback(galleryImagesLoadedCallback);
        super.onDestroy();
    }
}

File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/product/ProductDetailsActivity.kt
package com.mlsdev.mlsdevstore.presentaion.product

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.mlsdev.mlsdevstore.R
import com.mlsdev.mlsdevstore.databinding.ActivityProductDetailsBinding
import com.mlsdev.mlsdevstore.presentaion.BaseActivity
import com.mlsdev.mlsdevstore.presentaion.adapter.ProductGalleryAdapter
import com.mlsdev.mlsdevstore.presentaion.utils.ExtrasKeys
import javax.inject.Inject

class ProductDetailsActivity : BaseActivity() {
    @Inject
    lateinit var galleryAdapter: ProductGalleryAdapter
    lateinit var binding: ActivityProductDetailsBinding
    lateinit var viewModel: ProductDetailsViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_product_details)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(ProductDetailsViewModel::class.java)
        lifecycle.addObserver(viewModel)
        binding.viewModel = viewModel
        binding.productImagesGallery.adapter = galleryAdapter
        viewModel.setProductDetailsData(intent.extras)

        viewModel.imagesLiveData.observe(this, Observer { galleryAdapter.setImages(it) })
        viewModel.descriptionStateLiveData.observe(this, Observer {
            if (it) {
                val data = Bundle().apply { putString(ExtrasKeys.KEY_PRODUCT_DESCRIPTION, viewModel.description.get()) }
                val bottomSheet = DescriptionBottomSheetFragment()
                bottomSheet.arguments = data
                bottomSheet.show(supportFragmentManager, bottomSheet.tag)
            }
        })

        initErrorHandler()
        initToolbar(binding.toolbar)
        displayBackArrow(true)
    }

    private fun initErrorHandler() {
        errorInViewHandler.observeNetworkError(this, viewModel.networkErrorLiveData)
        errorInViewHandler.observeCommonError(this, viewModel.commonErrorLiveData)
        errorInViewHandler.observeTechError(this, viewModel.technicalErrorLiveData)
        errorInViewHandler.observeAuthError(this, viewModel.authErrorLiveData)
    }

    override fun onDestroy() {
        super.onDestroy()
        lifecycle.removeObserver(viewModel)
    }

    companion object {

        fun launch(context: Context, bundleProductData: Bundle) {
            context.startActivity(Intent(context, ProductDetailsActivity::class.java).apply { putExtras(bundleProductData) })
        }
    }
}
